tools/SetupFlow/DevHome.SetupFlow/ViewModels/Environments/EnvironmentCreationOptionsViewModel.cs

```diff
@@ -0,0 +1,269 @@
+
+        // Register to receive the EnvironmentCreationOptionsViewRequestMessage so that we can populate the configure environment page with the current
+        // adaptive card information. This handles the case where the view is loaded after the view model has finished loading the adaptive card.
+        WeakReferenceMessenger.Default.Register<EnvironmentCreationOptionsViewModel, CreationOptionsViewPageRequestMessage>(this, OnEnvironmentOptionsViewRequest);
+
+        WeakReferenceMessenger.Default.Register<EnvironmentCreationOptionsViewModel, CreationOptionsReviewPageDataRequestMessage>(this, OnReviewPageViewRequest);
+    }
+
+    /// <summary>
+    /// Weak reference message handler for when the selected provider changes in the select environment provider page. This will be triggered when the user
+    /// selects an item in the Select Environment Provider page. The next time the user goes to this configure environments page, we'll update the UI
+    /// with an adaptive card from the newly selected provider.
+    /// </summary>
+    /// <param name="message">Message data that contains the new provider details.</param>
+    public void Receive(CreationProviderChangedMessage message)
+    {
+        UpcomingProviderDetails = message.Value;
+    }
+
+    private void OnEndSetupFlow(object sender, EventArgs e)
+    {
+        ResetAdaptiveCardConfiguration();
+        WeakReferenceMessenger.Default.UnregisterAll(this);
+        _setupFlowViewModel.EndSetupFlow -= OnEndSetupFlow;
+    }
+
+    protected async override Task OnFirstNavigateToAsync()
+    {
+        // Does nothing, but we need to override this as the base expects a task to be returned.
+        await Task.CompletedTask;
+        AdaptiveCardRenderer = GetAdaptiveCardRenderer();
+    }
+
+    /// <summary>
+    /// Make sure we only get the list of ComputeSystems from the ComputeSystemManager once when the page is first navigated to.
+    /// All other times will be through the use of the sync button.
+    /// </summary>
+    protected async override Task OnEachNavigateToAsync()
+    {
+        // This doesn't do any awaitable work, but we're overriding the method to ensure that it's called.
+        // when we navigate to this page in the setup flow.
+        await Task.CompletedTask;
+        var curSelectedProviderId = CurProviderDetails?.ComputeSystemProvider?.Id ?? string.Empty;
+        var upcomingSelectedProviderId = UpcomingProviderDetails?.ComputeSystemProvider?.Id;
+
+        // Selected compute system provider  may havechanged so we need to update the adaptive card in the UI
+        // with new a adaptive card from the new provider.
+        CurProviderDetails = UpcomingProviderDetails;
+
+        IsAdaptiveCardSessionLoaded = false;
+
+        // Its possible that an extension could take a long time to load the adaptive card session.
+        // So we run this on a background thread to prevent the UI from freezing.
+        _ = Task.Run(() =>
```

Comment:
- Do we need to lock here?
- Are there any out of proc objects from extensions (e.g. providers) that we need to try catch?

Reply:
We've created a wrapper class for the compute system provider which try catches on method calls. See: https://github.com/microsoft/devhome/blob/c06b950560e2df63b555ffd06c8bebd093a11d62/common/Environments/Models/ComputeSystemProvider.cs#L46

I didn't think a lock would be needed as this method will go on to create a new extension adaptive card session, a new extension adaptive card, and the listeners will receive the new rendered adaptive card. I don't have concerns about synchronization, as each page only has one adaptive card for that particular page.